Add ca_certs argument for oauth and dropbox client

stanislau-arkhipenka opened the pull request:
<!--
Thank you for your pull request. Please provide a description below.
-->
This change adds ca_certs str argument for oauth and dropbox client classes. It may be used  with applications that packed into executable archives (like python par or py2exe). It also may be useful with some security related scenarios.

## **Checklist**
<!-- For completed items, change [ ] to [x]. -->

**General Contributing**
- [x] Have you read the Code of Conduct and signed the [CLA](https://opensource.dropbox.com/cla/)?

**Is This a Code Change?**
- [x] Non-code related change (markdown/git settings etc)
- [x] SDK Code Change
- [x] Example/Test Code Change

**Validation**
- [x] Does `tox` pass?
- [x] Do the tests pass? (no integration tests)

Commit:
Add ca_certs argument for oauth and dropbox client

greg-db commented:
Thanks for putting this together! I'll ask the team to review this, though I can't promise when that would be.

stanislau-arkhipenka commented:
Hey @greg-db , thanks for reply. May be you have any updates?

greg-db commented:
@stanislau-arkhipenka This is still open with the team, but I don't have an update on it.

rogebrd commented:
Hello @stanislau-arkhipenka, I took at look at the PR and code wise, it looks good to me. I want to get a better understanding for your use case before we make the decision to merge this. Could you help me understand why this is needed?

april commented on dropbox/session.py:
```diff
@@ -32,22 +32,33 @@
+    def __init__(self, *args, **kwargs):
+        self._ca_certs = kwargs.pop("ca_certs", None) or _TRUSTED_CERT_FILE
+        if not self._ca_certs:
+            raise FileNotFoundError("CA certificate not found")
```
This error is a bit misleading - it's not checking for the presence of a readable file (although maybe it should be?) - it's simply checking to see if `ca_certs` or `_TRUSTED_CERT_FILE` is set.

stanislau-arkhipenka commented:
Hello @rogebrd , 
Sorry for quite a long delay. 
There are two possible scenarios when this code may be useful:
1) In a secure environment with https reverse proxy (effectively corporate term for "man in the middle"). This diff will allows to use custom certificate issued by company.
2) (my case) For cases when developed application is packed into executable archive like [par](https://google.github.io/subpar/subpar.html) or [py2exe](https://pypi.org/project/py2exe/) (issues mentioned [here](https://www.dropboxforum.com/t5/Dropbox-API-Support-Feedback/py2exe/td-p/143992) and [here](https://stackoverflow.com/questions/29198947/program-wont-work-after-being-compiled-with-py2exe)).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dropbox/dropbox-sdk-python?pullRequest=385) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/dropbox/dropbox-sdk-python?pullRequest=385) before we can accept your contribution.<br/>**0** out of **3** committers have signed the CLA.<br/><br/>:x: Brent1LT<br/>:x: stanislau-arkhipenka<br/>:x: Stanislau Arkhipenka<br/><hr/>**Stanislau Arkhipenka** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dropbox/dropbox-sdk-python?pullRequest=385) it.</sub>

stanislau-arkhipenka commented:
> [![CLA assistant check](https://camo.githubusercontent.com/4c970efeed75dd01f6778bf7e33dec85e4171375cb69acb0927bfc9d437d78d9/68747470733a2f2f636c612d617373697374616e742e696f2f70756c6c2f62616467652f6e6f745f7369676e6564)](https://cla-assistant.io/dropbox/dropbox-sdk-python?pullRequest=385) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/dropbox/dropbox-sdk-python?pullRequest=385) before we can accept your contribution.
> **Stanislau Arkhipenka** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dropbox/dropbox-sdk-python?pullRequest=385) it.

Signed it

stanislau-arkhipenka commented:
Hey guys, any updates? I've signed CLA. Do you need anything else on my side?
Thanks!

greg-db commented:
Thanks for checking in. This is open with the team, but I don't have an update or timeline on it right now.

Brent1LT committed:
Merge branch 'main' into add_ca_certs_arg

Brent1LT commented:
Everything looks good to me but it looks like there are some integration tests that are failing before it can get merged in.

stanislau-arkhipenka committed:
Replace FileNotFoundError with AttributeError

New error better represents actual problem (CA not set)
And also compatible with python2.7

stanislau-arkhipenka commented:
> Everything looks good to me but it looks like there are some integration tests that are failing before it can get merged in.



> Everything looks good to me but it looks like there are some integration tests that are failing before it can get merged in.

Hey @Brent1LT , thanks for reply! 
Indeed there was a problem with FileNotFoundError (it doesn't supported in python2.7) - FIXED.
With regards to integration tests: I believe it is not related to this particular PR, as it seems persists across other open  PRs.
Based on error message my best guess is that some environmental variables on CI side are not populated correctly. I'm not sure, but it may be related to rename of envs from DROPBOX_* to SCOPED_* (btw same problem exists if you try to run tox -e test_integration )

stanislau-arkhipenka commented:
@rogebrd , @april , @Brent1LT 
How do you think can this PR be merged to upstream?
The Integration tests are failing because of missed environmental variables, and it doesn't relate to PR

dennissiemensma commented:
I've added a minor follow-up for this PR in #440 to prevent pickle from breaking.
